File: app/controllers/halo4/challenges_controller.rb
class Halo4::ChallengesController < Halo4::HomeController

	def index
		@challenge_data = H4Api.get_challenge_data()['Challenges']
		@challenge_categories = [ ]

		@challenge_data.each do |challenge|
			# check we no have category
			we_have = false
			@challenge_categories.map{ |c| we_have = true if (c['Id'] == challenge['CategoryId']) }

			if (!we_have)
				@challenge_categories << get_challenge_category_from_base_id(challenge['CategoryId'])
			end
		end
	end
end

File: Branch.App/Areas/Halo4/Controllers/ChallengesController.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using Branch.App.Areas.Halo4.Models;
using Branch.App.Filters;
using Branch.App.Helpers.Razor.Halo4;
using Branch.Models.Services.Halo4._343.DataModels;

namespace Branch.App.Areas.Halo4.Controllers
{
	public class ChallengesController : Controller
	{
		//
		// GET: /Halo4/Challenges/{slug}
		[ValidateH4ApiStatus]
		public ActionResult Index(string slug)
		{
			var challenges = GlobalStorage.H4WaypointManager.Challenges.Challenges;
			var challengeCategories = new List<MetadataModels.ChallengeCategory>();

			foreach (var challenge in from challenge in challenges
				let weHaveIt = (challengeCategories.FirstOrDefault(c => c.Id == (int) challenge.Category) != null)
				where !weHaveIt
				select challenge)
			{
				challengeCategories.Add(MetadataHelpers.GetChallengeCategory((int)challenge.Category));
			}
			var selectedCategory =
				challengeCategories.FirstOrDefault(c => c.Name.ToLowerInvariant().Replace(" ", "") == slug.ToLowerInvariant());

			return View(new ChallengesViewModel(selectedCategory, challengeCategories, challenges));
		}
	}
}
